src/mmda/predictors/heuristic_predictors/figure_table_predictors.py

```diff
@@ -110,58 +351,58 @@ def _predict(doc: Document, caption_type: str = 'Figure') -> List[SpanGroup]:
             doc (Document): document to make predictions on, it has to have fields layoutparser_span_groups and
             vila_span_groups
             caption_type (str): caption type to make prediction for can be Figure or Table
-        Returns: Returns list of spanGroups predictions
+        Returns: Returns dictionary with keys 'predictions', 'predictions_captions', 'predictions_relations'
 
         """
-        assert doc.layoutparser_span_groups
-        assert doc.vila_span_groups
-        if caption_type == 'Figure':
-            merged_boxes_caption_dict = FigureTablePredictions.merge_spans(doc.vila_span_groups)
-            merged_boxes_fig_dict = FigureTablePredictions._merge_boxes(doc.layoutparser_span_groups,
-                                                                        merged_boxes_caption_dict)
-        else:
-            merged_boxes_caption_dict = FigureTablePredictions.merge_spans(doc.vila_span_groups, caption_content='tab')
-            merged_boxes_fig_dict = FigureTablePredictions._merge_boxes(doc.layoutparser_span_groups,
-                                                                        merged_boxes_caption_dict, types=['Table'])
-
         predictions = []
-        for page in range(len(tqdm(doc.images))):
-            if merged_boxes_caption_dict.get(page) and merged_boxes_fig_dict.get(page):
+        predictions_captions = []
+        predictions_relations = []
+        for page in range(len(tqdm(self.doc.pages))):
+            if merged_boxes_caption_dict.get(page) and merged_boxes_fig_tab_dict.get(page):
                 cost_matrix = np.zeros(
-                    (len(merged_boxes_fig_dict[page]),
+                    (len(merged_boxes_fig_tab_dict[page]),
                      len(merged_boxes_caption_dict[page])))
-                for j, fig_box in enumerate(merged_boxes_fig_dict[page]):
-                    for i, caption_box in enumerate(merged_boxes_caption_dict[page]):
+                for j, fig_box in enumerate(merged_boxes_fig_tab_dict[page]):
+                    for i, span_group in enumerate(merged_boxes_caption_dict[page]):
+                        caption_box = span_group.box
                         assert hasattr(fig_box, 'box')
-                        assert hasattr(caption_box, 'box')
                         distance = FigureTablePredictions._get_object_caption_distance(
-                            fig_box.box, caption_box.box)
+                            fig_box.box, caption_box)
 
                         cost_matrix[j][i] = distance
                         if caption_type == 'Figure':
                             cost_matrix[j][i] = distance if distance > 0 else 900
 
                 row_ind, col_ind = linear_sum_assignment(cost_matrix)
                 for row, col in zip(row_ind, col_ind):
-                    predictions.append(SpanGroup(spans=[Span(
+                    span_group = SpanGroup(spans=[Span(
```

Comment:
This construction is slightly awkward because you are subsequently pulling the box group out and then adding it to a separate list. At the end of this process, we don't want the SG to have a BG anymore, so ideally we don't put it there in the first place. What about:

```python
for row, col in zip(row_ind, col_ind):
    span_group = SpanGroup(
        spans=[
            Span(
                start=merged_boxes_caption_dict[page][col].start,
                end=merged_boxes_caption_dict[page][col].box
            )
        ]
    )

    box_group = BoxGroup(boxes=[merged_boxes_fig_tab_dict[page][row][box])
```

Note I have done three things:

1. never put the box group on the SG
2. don't put boxes on the individual spans (we don't want/need them there, will break spp persistence and violates how Entities will be constructed in mmda refactor)
3. dropped Metadata from your SG because this was only being used to differentiate between table and figure captions. And you are not saving those as separate lists of spangroups, so they are disambiguated there instead.